### DISCUSSION TITLE
How do I make the Entities to not get detached when using nested Unis

### DISCUSSION THREAD

**Discussion by underscore05** [quarkusio/quarkus#44621] [Q&A] [answered]
I have this code where I try to get an entity in a sequential manner. At the last flatMap, I can see that agent and customer is still assigned to newJobPosting. But persisting it causes an error like `ava.util.concurrent.CompletionException: org.hibernate.PersistentObjectException: detached entity passed to persist`. I suspect that the agent and customer is already detached at that point but I don't know what to do to attached them again. I am using reactive-panache and repository for agent, customer and jobposting entity.

```java
Log.info("Creating new job-posting");
return Uni.createFrom().item(() -> modelMapper.map(request, JobPosting.class))
    .flatMap((newJobPosting) -> getAgentOrException(jwtId).map(agent -> {
        newJobPosting.setAgent(agent);
        return newJobPosting;
    }))
    .flatMap((newJobPosting) -> {
        if(request.getAssignedCustomerId() != null && JobPostingVisibility.PRIVATE_POST.equals(request.getVisibility())) {
            return getCustomerOrException(request.getAssignedCustomerId())
                    .map(customer -> {

                        newJobPosting.setCustomer(customer);
                        return newJobPosting;
                    });
        }
        return Uni.createFrom().item(newJobPosting);
    })
    .flatMap(newJobPosting -> {
        if(newJobPosting.getAgent() != null) {
            Log.infof("Has Agent! %s", newJobPosting.getAgent().getJwtId());
        }
        if(newJobPosting.getCustomer() != null) {
            Log.infof("Has Customer! %s", newJobPosting.getCustomer().getJwtId());
        }
        return jobPostRepository.persist(newJobPosting);
    });
```

**Comment by geoand** [comment]
cc @DavideD

**Comment by DavideD** [comment]
Without looking at the rest of the code and the mapping, I'm not sure what's the best solution.
Usually, it's better to try not to close the session too soon.

To reattach an object to the session:
* you can reload it
* or, you can merge it

Reloading it works if the object doesn't have any changes you want to save. the Hibernate session has several methods to do it. Otherwise, you can merge it using `session.merge(...)`.

I don't know why, but Panache doesn't seem to have a merge operator out of the box, but adding this to your repository should work:
```java
class JobPostRepository implements PanacheRepository<JobPost> {
        // ...

	public Uni<JobPost> merge(JobPost jobPost) {
		return Panache.getSession().chain( session -> session.merge( jobPost ) );
	}
}
```

You should be able to adapt this example to what you need.

I hope this help.

**Reply by underscore05** [reply]
I dont know when exactly I am closing the session. All the `getAgentOrException` and `getCustomerOrException` is doing is a `repository.findById`. The method that wraps this code is annotated with `@withTransaction` as well.

**Reply by underscore05** [reply]
Syntax wise. How can I do stuff like getEntity1, conditionally getEntity2, then assign entity1 and entity2 to entity3? In my example I use flatMap but unfortunately it looks like entity1 and entity2 got detached upon persisting entity3.

**Reply by DavideD** [reply]
But are you sure those are the entities that are causing the exception? Are the associations bidirecational?
What's the entity mapping of `JobPosting`?

If you provide a small test case, I can have a look. Otherwise, I don't think we have enough information.

**Reply by underscore05** [reply]
They are not bi-directional. Only the job-posting has a declaration for agent and customer. It is @manytoone relation to both agent and customer entity. I can provide one. I am just driving right now :) Thanks for answering.

**Reply by underscore05** [reply, marked as answer]
Hello @DavideD closing this discussion. 
Chaining the `flatMap` does not make the entity to get detached. 
It's a different issue all along. 
Changing `modelMapper.map(request, JobPosting.class)` to just `new JobPosting()` fixes my code. Thanks for your time.

**Reply by underscore05** [reply]
Just an update on what really happen. `ModelMapper` just fills-up the `id` field of my `JobPosting` entity and uses whatever `Long` field is available my `request` object regardless whether it matches the `id` name or not (in my case it uses `assignedCustomerId` and put it as `id` value of `JobPosting`). Setting my model-mapper's matching strategy to strict fixes it.
```
modelMapper.getConfiguration().setMatchingStrategy(MatchingStrategies.STRICT);
```
